### ISSUE TITLE
train on dt4d

### ISSUE THREAD

**Issue by Native537** [xieyizheng/hybridfmaps#1]
Hi,
     

      I always have the phenomenon that the loss is always NAN when I train on the dt4d dataset, I have adjusted the learning rate and it doesn't help, what should I do to fix it.

Best wish.

**Comment by xieyizheng** [comment]
Hi Native537,
Thank you for raising this issue. We have also encountered this phenomenon in our experiments. It likely indicates a bug in the current release of the code. We will investigate this and release a patch as soon as possible. 
Best regards

**Comment by xieyizheng** [comment]
Hi Native537, as we're revisiting the code, we found some other source of training instability and will try to improve that, but haven't met the nan issue yet. at the same time would you post as much your steps of reproduction and the training logs, those information could be greatly helpful and appreciated!

**Comment by Native537** [comment]
Hi Xie，

Firstly, my environment is as follows, torch=1.12.1+cu113 torchvision=0.13.1, torchaudio=0.12.1.While training on dt4d，at the first epoch, iteration is 100 ，there is a loss of NAN,and no matter how many adjustments are made, it's the same thing every time.
I'm very sorry, the training logs was cleaned up by me, I'll get back to reproducing it and making it available as soon as I can. Lastly, I would like to ask if there is any possibility that it is related to my environment.
Best regards

**Comment by xieyizheng** [comment]
Hi thanks for the provided info! That's very helpful. Yes, the environment could be an issue for incompatibilities, you could give it a try the one we've provided. In general, it's also recommended to try out a few different random seeds even with the same configs to "outrun" randomness, it would get slightly different result each time and sometimes can get away with tiny issues just by luck. 
To disclose some extra context for the NaN situation here, one of the sources we've met comes from the numerical instability of the elastic basis, we should already have some implementation in-place to handle those, for us we're simply replacing those entries with nearest non-nan values. But there could also be some other sources we haven't uncovered yet.

**Comment by YeTianwei** [comment]
Hi Xie ,

I've encountered the same problem while training on the DT4D dataset where all loss values consistently turn into NaN starting exactly at iteration 4280. I have tried making adjustments, but the issue persists.

Environment Details:

- PyTorch version: torch==2.2.2+cu121, using the `environment.yml` you provided.
- Training configuration: Using the exact settings provided in your `options/hybrid_ulrssm/train/dt4d.yaml` file.
- GPU: single TITAN TRX.

I would appreciate any guidance or suggestions on how to debug and resolve this NaN loss issue.

Best regards.

**Comment by xieyizheng** [comment]
Hi thanks for raising this, I can try to take a look again at some point, although life has been a bit busy. But I encourage PRs if you or someone can come up with a elegant workaround


At least , it looks like the issue is quite consistent and reproducible and mostly likely can be easily fixed

**Comment by xieyizheng** [comment]
I suspect this as a numerical instability thing,even though not exactly sure where. But a very ugly fix might be to just detect and rollback the network weights and try again or move on training iteration

**Comment by YeTianwei** [comment]
Hi, I've managed to solve the NaN loss issue. As you said, the problem stems from numerical instability, which I address with two little changes within the `ExpandedResolventFMNet` class:

- Replace the explicit Kronecker product with an Einstein summation
```python
def efficient_kron(a, b):
    s_a, s_b = a.size(), b.size()
    return torch.einsum('ab,cd->acbd', a, b).reshape(s_a[0]*s_b[0], s_a[1]*s_b[1])
At_Ik = efficient_kron(A_t, Ik) # torch.kron(A_t, Ik)
```
- Add regularization when solving for C
```python
op = first + self.lmbda * second
op_reg = op + 1e-6 * torch.eye(op.size(0), device=op.device)
C = torch.linalg.solve(op_reg, rhs)
```
I hope this information is helpful for others encountering similar issues.

**Comment by xieyizheng** [comment]
Awesome! thanks a lot for sharing! feel free to open a pull request as well, in case someone else is able to try it out and confirm, I'd be happy to merge it
